libctru/source/services/mic.c

```diff
@@ -1,271 +1,226 @@
+Result MICU_StartSampling(MIC_Encoding encoding, MIC_SampleRate sampleRate, u32 offset, u32 size, bool loop)
 {
-	Result ret=0;
-	u32 *cmdbuf = getThreadCommandBuffer();
-
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
 	cmdbuf[0] = IPC_MakeHeader(0x3,5,0); // 0x30140
-	cmdbuf[1] = unk0;
-	cmdbuf[2] = unk1;
-	cmdbuf[3] = sharedmem_baseoffset;
-	cmdbuf[4] = sharedmem_endoffset;
-	cmdbuf[5] = unk2;
+	cmdbuf[1] = encoding;
+	cmdbuf[2] = sampleRate;
+	cmdbuf[3] = offset;
+	cmdbuf[4] = size;
+	cmdbuf[5] = loop;
+
+	if (R_FAILED(ret = svcSendSyncRequest(micHandle)) != 0) return ret;
+	return cmdbuf[1];
+}
 
-	if((ret = svcSendSyncRequest(MIC_handle))!=0)return ret;
+Result MICU_AdjustSampling(MIC_SampleRate sampleRate)
+{
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
+	cmdbuf[0] = IPC_MakeHeader(0x4,1,0); // 0x40040
+	cmdbuf[1] = sampleRate;
 
-	return (Result)cmdbuf[1];
+	if (R_FAILED(ret = svcSendSyncRequest(micHandle)) != 0) return ret;
+	return cmdbuf[1];
 }
 
-Result MIC_cmd5(void)
+Result MICU_StopSampling(void)
 {
-	Result ret=0;
-	u32 *cmdbuf = getThreadCommandBuffer();
-
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
 	cmdbuf[0] = IPC_MakeHeader(0x5,0,0); // 0x50000
 
-	if((ret = svcSendSyncRequest(MIC_handle))!=0)return ret;
-
-	return (Result)cmdbuf[1];
+	if (R_FAILED(ret = svcSendSyncRequest(micHandle)) != 0) return ret;
+	return cmdbuf[1];
 }
 
-Result MIC_GetCNTBit15(u8 *out)
+Result MICU_IsSampling(bool* sampling)
 {
-	Result ret=0;
-	u32 *cmdbuf = getThreadCommandBuffer();
-
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
 	cmdbuf[0] = IPC_MakeHeader(0x6,0,0); // 0x60000
 
-	if((ret = svcSendSyncRequest(MIC_handle))!=0)return ret;
-
-	if(out)*out = cmdbuf[2];
-
-	return (Result)cmdbuf[1];
+	if (R_FAILED(ret = svcSendSyncRequest(micHandle)) != 0) return ret;
+	if (sampling) *sampling = cmdbuf[2];
+	return cmdbuf[1];
 }
 
-Result MIC_GetEventHandle(Handle *handle)
+Result MICU_GetEventHandle(Handle* handle)
 {
-	Result ret=0;
-	u32 *cmdbuf = getThreadCommandBuffer();
-
-	if(MIC_event)
-	{
-		*handle = MIC_event;
-		return 0;
-	}
-
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
 	cmdbuf[0] = IPC_MakeHeader(0x7,0,0); // 0x70000
 
-	if((ret = svcSendSyncRequest(MIC_handle))!=0)return ret;
-
-	if(handle)*handle = cmdbuf[3];
-
-	return (Result)cmdbuf[1];
+	if (R_FAILED(ret = svcSendSyncRequest(micHandle)) != 0) return ret;
+	if (handle) *handle = cmdbuf[3];
+	return cmdbuf[1];
 }
 
-Result MIC_SetControl(u8 value)
+Result MICU_SetGain(u8 gain)
 {
-	Result ret=0;
-	u32 *cmdbuf = getThreadCommandBuffer();
-
+	Result ret = 0;
+	u32* cmdbuf = getThreadCommandBuffer();
 	cmdbuf[0] = IPC_MakeHeader(0x8,1,0); // 0x80040
-	cmdbuf[1] = value;
+	cmdbuf[1] = gain;
 
-	if((ret = svcSendSyncRequest(MIC_handle))!=0)return ret;
-
-	return (Result)cmdbuf[1];
+	if (R_FAILED(ret = svcSendSyncRequest(micHandle)) != 0) return ret;
```

Comment:
Should have been `if (R_FAILED(ret = svcSendSyncRequest(micHandle)))` (i.e. remove the != 0). This happens in other functions too.